frame/system/src/lib.rs

```diff
@@ -614,12 +643,24 @@ decl_module! {
 
 		/// Kill all storage items with a key that starts with the given prefix.
 		///
+		/// **NOTE:** We rely on the Root origin to provide us the number of subkeys under
+		/// the prefix we are removing to accurately calculate the weight of this function.
+		///
 		/// # <weight>
 		/// - `O(P)` where `P` amount of keys with prefix `prefix`
 		/// - `P` storage deletions.
+		/// - Base Weight: 0.834 * P µs
+		/// - Writes: Number of subkeys + 1
 		/// # </weight>
-		#[weight = (0, DispatchClass::Operational)]
-		fn kill_prefix(origin, prefix: Key) {
+		#[weight = FunctionOf(
+			|(_, &subkeys): (&Key, &u32)| {
+				T::DbWeight::get().writes(Weight::from(subkeys) + 1)
```

Comment:
I wonder if this is not too big estimate. in case you remove a prefix you only have to change on path in the trie.
For the root calculation you don't have to update the nodes in N path but just one path.
But I think it is fine for now.

Reply:
This is actually what I originally thought too, but we had this discussion: https://github.com/paritytech/substrate/pull/5593#discussion_r406728786

Let me clarify